### ISSUE TITLE
Core dump when request forwarder receiving data 

### ISSUE THREAD

**Issue by JackyWoo** [JDRaftKeeper/RaftKeeper#334]
### RaftKeeper version
v2.1.0

### Analysis
Let's open the core dump file and show the stack trace.

lldb -c core-ReqFwdRecv#43-1384962-1009-1720544680 tmp/raftkeeper

(lldb) bt

```
* thread #1, name = 'raftkeeper', stop reason = signal SIGSEGV
  * frame #0: 0x00000000006fe12c raftkeeper`RK::RequestForwarder::runReceive(unsigned long) [inlined] std::__1::shared_ptr<RK::ForwardConnection>::shared_ptr(this=<unavailable>, __r=<unavailable>) at memory:3097:18
    frame #1: 0x00000000006fe12c raftkeeper`RK::RequestForwarder::runReceive(unsigned long) [inlined] std::__1::shared_ptr<RK::ForwardConnection>::operator=(this=0x00007fb53f3a9fd0, __r=<unavailable>) at memory:3220:5
    frame #2: 0x00000000006fe12c raftkeeper`RK::RequestForwarder::runReceive(this=0x00007fb58d443408, runner_id=7) at RequestForwarder.cpp:167:32
    frame #3: 0x000000000063cc3e raftkeeper`ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) [inlined] std::__1::__function::__policy_func<void ()>::operator()(this=0x00007fb53f3aa100) const at functional:2221:16
    frame #4: 0x000000000063cc35 raftkeeper`ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) [inlined] std::__1::function<void ()>::operator()(this=0x00007fb53f3aa100) const at functional:2560:12
    frame #5: 0x000000000063cc35 raftkeeper`ThreadPoolImpl<ThreadFromGlobalPool>::worker(this=0x00007fb58c853998, thread_it=std::__1::list<ThreadFromGlobalPool, std::__1::allocator<ThreadFromGlobalPool> >::iterator @ 0x00007fb53f3aa0f8) at ThreadPool.cpp:265:17
```

(lldb) frame select 2

```
frame #2: 0x00000000006fe12c raftkeeper`RK::RequestForwarder::runReceive(this=0x00007fb58d443408, runner_id=7) at RequestForwarder.cpp:167:32
   164 	                ptr<ForwardConnection> connection;
   165 	                {
   166 	                    std::lock_guard<std::mutex> lock(connections_mutex);
-> 167 	                    connection = connections[leader][runner_id];
   168 	                }
   169 	
   170 	                if (connection && connection->isConnected())
```

(lldb) frame variable leader

```
(int32_t) leader = 3
```

(lldb) expr ((RK::RequestForwarder*)0x00007fb58d443408)->connections

```
(std::unordered_map<unsigned int, std::vector<std::shared_ptr<RK::ForwardConnection> > >) $0 = size=3 {
  [0] = {
    first = 3
    second = size=0 {}
  }
  [1] = {
    first = 1
    second = size=48 {
      [0] = std::__1::shared_ptr<RK::ForwardConnection>::element_type @ 0x00007fb58c705798 strong=1 weak=1 {
        __ptr_ = 0x00007fb58c705798
      }
      [1] = std::__1::shared_ptr<RK::ForwardConnection>::element_type @ 0x00007fb58c705958 strong=1 weak=1 {
        __ptr_ = 0x00007fb58c705958
      }
      ...
  [2] = {
    first = 2
    second = size=48 {
      [0] = std::__1::shared_ptr<RK::ForwardConnection>::element_type @ 0x00007fb58c700398 strong=1 weak=1 {
        __ptr_ = 0x00007fb58c700398
      }
      [1] = std::__1::shared_ptr<RK::ForwardConnection>::element_type @ 0x00007fb58c700558 strong=1 weak=1 {
        __ptr_ = 0x00007fb58c700558
      }
      [2] = std::__1::shared_ptr<RK::ForwardConnection>::element_type @ 0x00007fb58c700718 strong=1 weak=1 {
        __ptr_ = 0x00007fb58c700718
      }
      ...
```

Found that there are 3 servers in `connections` and  `connections` for server 3 is empty. There must be a logical error for the cluster has 3 servers, so `connections` should have 2 server connections.

Let's confirm which server id it is.

(lldb) expr this->server
```
(std::shared_ptr<RK::KeeperServer>) $9 = std::__1::shared_ptr<RK::KeeperServer>::element_type @ 0x00007fb58c7001d8 strong=4 weak=1 {
  __ptr_ = 0x00007fb58c7001d8
}
```

(lldb) expr ((RK::KeeperServer*)0x00007fb58c7001d8)
```
(RK::KeeperServer *) $10 = 0x00007fb58c7001d8
```
(lldb) expr ((RK::KeeperServer*)0x00007fb58c7001d8)->my_id
```
(int32_t) $11 = 3
```

Now e know the server id is 3, because it should not foward request it to  itself. so `connections` should not contains server 3.

Let's introspect `raft_server`

```
(lldb) expr ((RK::KeeperServer*)0x00007fb58c7001d8)->raft_instance
(nuraft::ptr<nuraft::raft_server>) $12 = std::__1::shared_ptr<nuraft::raft_server>::element_type @ 0x00007fb58c849018 strong=6 weak=2 {
  __ptr_ = 0x00007fb58c849018
}

(lldb) expr ((nuraft::raft_server*)0x00007fb58c849018)
(nuraft::raft_server *) $13 = 0x00007fb58c849018

(lldb) expr ((nuraft::raft_server*)0x00007fb58c849018)->leader_

(std::atomic<int>) $14 = {
  Value = 3
}

(lldb) expr ((nuraft::raft_server*)0x00007fb58c849018)->role_
(std::atomic<nuraft::srv_role>) $15 = {
  Value = 3  ///  3 is LEADER
}
```

Found that `role_` is LEADER and `leader_` is 3.

Let's check code in `RequestForwarder::runReceive`

```
            if (!server->isLeader() && server->isLeaderAlive())
            {
                int32_t leader = server->getLeader();
                ptr<ForwardConnection> connection;
                {
                    std::lock_guard<std::mutex> lock(connections_mutex);
                    connection = connections[leader][runner_id];
                }
```

In summary `server->isLeader()` returns false, but when run at line `int32_t leader = server->getLeader()`  it is leader which is caused by leader switch.

It is a data condition problem we should check leader(can not be myself) before get connection.
